flink-python/pyflink/fn_execution/datastream/window/window_assigner.py

```diff
@@ -0,0 +1,376 @@
+        return windows
+
+    def get_default_trigger(self, env) -> Trigger[T, W]:
+        if self._is_event_time is True:
+            return EventTimeTrigger()
+        else:
+            return ProcessingTimeTrigger()
+
+    def get_window_serializer(self) -> TypeSerializer[W]:
+        return TimeWindowSerializer()
+
+    def is_event_time(self) -> bool:
+        return self._is_event_time
+
+    def __repr__(self) -> str:
+        return "SlidingWindowAssigner(%s, %s, %s, %s)" % (self._size, self._slide, self._offset, self._is_event_time)
+
+
+class CountSlidingWindowAssigner(WindowAssigner[object, CountWindow]):
+    """
+    A WindowAssigner that windows elements into sliding windows based on the count number of
+    the elements. Windows can possibly overlap.
+    """
+    def __init__(self, window_size: int, window_slide: int):
+        """
+        Windows this KeyedStream into sliding count windows.
+        :param window_size: The size of the windows in number of elements.
+        :param window_slide: The slide interval in number of elements.
+        """
+        self._window_size = window_size
+        self._window_slide = window_slide
+        self._count = None  # type: ValueState
+        self._counter_state_descriptor = ReducingStateDescriptor(
+            "slide-count-assigner", lambda a, b: a + b, Types.BIG_INT())
+
+    def assign_windows(self,
+                       element: T,
+                       timestamp: int,
+                       context: 'WindowAssigner.WindowAssignerContext') -> Collection[W]:
+        count_descriptor = ValueStateDescriptor('slide-count-assigner', Types.LONG())
+        self._count = context.get_runtime_context().get_state(count_descriptor)
+        count_value = self._count.value()
+        if count_value is None:
+            current_count = 0
+        else:
+            current_count = count_value
+        self._count.update(current_count + 1)
+        last_id = current_count // self._window_slide
+        last_start = last_id * self._window_slide
+        last_end = last_start + self._window_size - 1
+        windows = []
+        while last_id >= 0 and last_start <= current_count <= last_end:
+            if last_start <= current_count <= last_end:
+                windows.append(CountWindow(last_id))
+            last_id -= 1
+            last_start -= self._window_slide
+            last_end -= self._window_slide
+        return windows
+
+    def get_default_trigger(self, env) -> Trigger[T, W]:
+        return CountTrigger(self._window_size)
+
+    def get_window_serializer(self) -> TypeSerializer[W]:
+        return CountWindowSerializer()
+
+    def is_event_time(self) -> bool:
+        return False
+
+    def __repr__(self):
+        return "CountSlidingWindowAssigner(%s, %s)" % (self._window_size, self._window_slide)
+
+
+class SessionWindowAssigner(MergingWindowAssigner[object, TimeWindow]):
```

Comment:
It has already support dynamic session windows in the Java API and it would be great to align with it. See DynamicEventTimeSessionWindows for more details.